### ISSUE TITLE
Autocomplete characters should be configurable

### ISSUE THREAD

**Issue by cbrnr** [Microsoft/vscode#38352]
I've mentioned this in #15899, but this should really be a new issue. Just like Atom, Code should have an option to configure autocomplete characters. Currently, only brackets are auto-completed (and wrapped around selections), but it would be really useful if this also worked with quotes and backticks (especially in Markdown documents).

Here's the relevant settings page in Atom:
![screen shot 2017-11-14 at 22 26 26](https://user-images.githubusercontent.com/4377312/32806059-9fd20298-c98b-11e7-9cca-ce5010288eda.png)

**Comment by alexandrudima** [comment]
Today, this is something that is contributed via language configuration. Each language defines different sets of auto-closing brackets. We had them implemented since before we had language specific settings implemented...

:+1: This should be user-configurable

**Comment by cbrnr** [comment]
But so far quotes are not supported right? I just tried with a Markdown document and only brackets work.

**Comment by alexandrudima** [comment]
Markdown: https://github.com/Microsoft/vscode/blob/c7c988c5af378cf4fb4de62066f3783fb5f2e57b/extensions/markdown/language-configuration.json#L15

Each language defines what auto-closing pairs it supports based on its syntax.

**Comment by cbrnr** [comment]
Nice, but backticks don't work for me. If I type in a backtick, the second closing backtick doesn't appear. Similarly, if I select a word and type a backtick, the word is not surrounded by backticks but actually is overwritten by the tick.

**Comment by alexandrudima** [comment]
@cbrnr Yes, backticks (`` ` ``) and quotes (`` ' ``, `` " ``) do not auto close in markdown because markdown defines only the following auto-closing pairs: `{}`, `[]`, `()`, `<>`. A developer looked at the markdown syntax and decided those are the best set of auto-closing pairs given the Markdown syntax. That probably does not suite every need, hence this feature request, that end-users should be able to customize them. I agree.

https://github.com/Microsoft/vscode/blob/c7c988c5af378cf4fb4de62066f3783fb5f2e57b/extensions/markdown/language-configuration.json#L15

```
"autoClosingPairs": [
		{
			"open": "{",
			"close": "}"
		},
		{
			"open": "[",
			"close": "]"
		},
		{
			"open": "(",
			"close": ")"
		},
		{
			"open": "<",
			"close": ">",
			"notIn": [
				"string"
			]
		}
],
```

**Comment by Serhioromano** [comment]
All those are simple one char closing pairs. What if I have pair like this `{"open": "VAR_INPUT", "close": "END_VAR"}`. How would I enter it in settings? `VAR_INPUTEND_VAR` or with whitespace `VAR_INPUT END_VAR`, Both have problems. In first you do not know where pair separates. In second it is not like `{}` that is entered without space.

Second, my topic #57838 was closed to be tracked here, but my request is completely different thing. In my extension language support more than 30 pairs. And every pair like 10 to 40 symbols. I asked to make every pair configurable through extension settings that would be simple. But to retype all those pairs in configuration value as tex, would be a nightmare.

**Comment by coding2012** [comment]
Please fix this issue:
https://stackoverflow.com/questions/51311707/disable-auto-closing-for-html-comments-in-vs-code

Thanks!

**Comment by irvnriir** [comment]
anyone can edit `%directories%\Microsoft VS Code\resources\app\extensions\%theLanguage%\language-configuration.json` . but yes, they should be configurable trough settings (settings.json at least) .

it allow setting up multiple-characters units, like this line from Python `{ "open": "r\"", "close": "\"", "notIn": ["string", "comment"] },` .

**Comment by plwalsh** [comment]
@alexdima If the language configuration for Rust does not include angle brackets as part of its `autoClosingPairs`, how come I'm seeing a closing bracket appear when I type an opening bracket in my `*.rs` files, with the status bar reporting the language as `Rust`? It happens when trying to write a "less than" conditional if-statement.

https://github.com/microsoft/vscode/blob/74fdb8ec08591e5c11e0dcd6e8b3a613e1755051/extensions/rust/language-configuration.json#L11-L16

**Comment by User087** [comment]
> @cbrnr Yes, backticks (`` ` ``) and quotes (`'`, `"`) do not auto close in markdown because markdown defines only the following auto-closing pairs: `{}`, `[]`, `()`, `<>`. A developer looked at the markdown syntax and decided those are the best set of auto-closing pairs given the Markdown syntax. That probably does not suite every need, hence this feature request, that end-users should be able to customize them. I agree.
> 
> https://github.com/Microsoft/vscode/blob/c7c988c5af378cf4fb4de62066f3783fb5f2e57b/extensions/markdown/language-configuration.json#L15
> 
> ```
> "autoClosingPairs": [
> 		{
> 			"open": "{",
> 			"close": "}"
> 		},
> 		{
> 			"open": "[",
> 			"close": "]"
> 		},
> 		{
> 			"open": "(",
> 			"close": ")"
> 		},
> 		{
> 			"open": "<",
> 			"close": ">",
> 			"notIn": [
> 				"string"
> 			]
> 		}
> ],
> ```

backticks issue seems to be fixed by #184532, but the ability for users to customise is still desirable.

**Comment by alexrecuenco** [comment]
Any updates on this? It seems a bit abandoned, but it would be great.

**Comment by jarmoza** [comment]
Re: https://github.com/microsoft/vscode/issues/61854

Yes, +1 for adding this as a setting easily configurable by users.

**Comment by unkowncoder01** [comment]
> but the ability for users to customise is still desirable.

same I'd like to be able customize auto-close desired characters for desired languages. Not just markdown, I'd like also for Plain Text, currently the auto-close quotes doesn't work for Plain Text:

![Image](https://github.com/user-attachments/assets/3f4adedd-add6-44b2-a073-201aacbfeecf)

Plus, I wish there's a customizable hotkey that can be triggered after completing an auto-close which places the cursor/text cursor's position outside the auto-closed characters with a space to type the next word immediately, example:

from `"Hello|"` to `"Hello" |`

yes I've heard about the extension TabOut but it has limitations: limited to Tab key only + the cursor position does place outside the auto-closed characters but without a space, which forces the user to click space

**Comment by unkowncoder01** [comment]
> Markdown: https://github.com/Microsoft/vscode/blob/c7c988c5af378cf4fb4de62066f3783fb5f2e57b/extensions/markdown/language-configuration.json#L15
> 
> Each language defines what auto-closing pairs it supports based on its syntax.

@alexdima is there also a language-configuration.json for Plain Text? I cant find it in resources\app\extensions\

**Comment by gabuz0me** [comment]
Do we have any update on this feature request?

**Comment by EXealous** [comment]
Really really need this to work for mermaid (*.mmd) files.
